Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/qrcode/scan/SubmissionQRCodeScanViewModel.kt

```diff
@@ -56,10 +56,11 @@ class SubmissionQRCodeScanViewModel @AssistedInject constructor(
         try {
             registrationState.postValue(RegistrationState(ApiRequestState.STARTED))
             val testResult = submissionRepository.asyncRegisterDeviceViaGUID(scanResult.guid!!)
+            // Order here is important.to void race condition the coroutine will get cancelled
+            // when registrationState.postValue() is called before
+            testResultDataCollector.saveTestResultAnalyticsSettings(testResult)
```

Comment:
What is the reason that you did not choose to grab the test result in the SubmissionRepository? No issues with race conditions there.

Reply:
Yes, it should not `SubmissionRepository` is singleton. As TestResult metadata should be collected only for QR Code I did not want to add a new dependency that will live in the repo object as long as it is there. `SubmissionRepository` is used in other places too.

Reply:
Adding it to the ViewModel, it will serve its purpose for the use case and it is gone once the user is not on this screen